testsuite/tests/typing-gadts/pr5985.ml

```diff
@@ -54,17 +56,19 @@ let magic (x : int) : bool  =
 Line 1, characters 0-49:
 1 | type 'x t = A of 'a constraint 'x = [< `X of 'a ] ;; (* fail *)
     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
-Error: In this definition, a type variable cannot be deduced
-       from the type parameters.
+Error: In the expanded definition
+         type 'b t = A of 'a constraint 'b = [< `X of 'a ]
+       the type variable 'a cannot be deduced from the type parameters.
```

Comment:
I wonder if it would be possible to ask the type-checker to reuse the user variable name `'x` here. (The `Ctype.newvar*` combinators allow to provide name hints, but I don't know if they are being used directly here.)

Reply:
Sadly, I wasn't able to find the spot where to change this after looking a bit

Reply:
I did a little work to figure out what's happening, and here's my conclusion (it was very hard to follow, so this is probably wrong):

When we have a constraint, at some point, we drop the explicit representation of a constraint (lhs and rhs of constraint), and encode it directly in the type parameter. In this case, the type parameter is `[< `X of 'a ]` (I verified this by printing the type parameters). So the naming of the type variables for type parameters seems to be explicitly lost.

(PS: I definitely appreciate the PRs, such as #11936, to add more comments to the type checker, even just overview comments to explain what's going on. I think most of the time, I'm very lost when I start looking for how something works.)